js/packages/openinference-instrumentation-langchain/src/utils.ts

```diff
@@ -0,0 +1,299 @@
+    const value = attributes[key];
+
+    if (value == null) {
+      continue;
+    }
+
+    if (isObject(value)) {
+      Object.assign(result, flattenAttributes(value, newKey));
+    } else if (Array.isArray(value)) {
+      value.forEach((item, index) => {
+        if (isObject(item)) {
+          Object.assign(result, flattenAttributes(item, `${newKey}.${index}`));
+        } else {
+          result[`${newKey}.${index}`] = item;
+        }
+      });
+    } else if (isAttributeValue(value)) {
+      result[newKey] = value;
+    }
+  }
+  return result;
+}
+
+function getOpenInferenceSpanKindFromRunType(runType: string) {
+  const normalizedRunType = runType.toUpperCase();
+  if (normalizedRunType.includes("AGENT")) {
+    return OpenInferenceSpanKind.AGENT;
+  }
+
+  if (normalizedRunType in OpenInferenceSpanKind) {
+    return OpenInferenceSpanKind[
+      normalizedRunType as keyof typeof OpenInferenceSpanKind
+    ];
+  }
+  return "UNKNOWN";
+}
+
+function formatIO({
+  io,
+  type,
+}: {
+  io: Run["inputs"] | Run["outputs"];
+  type: "input" | "output";
+}) {
+  let valueAttribute: string;
+  let mimeTypeAttribute: string;
+  switch (type) {
+    case "input": {
+      valueAttribute = SemanticConventions.INPUT_VALUE;
+      mimeTypeAttribute = SemanticConventions.INPUT_MIME_TYPE;
+      break;
+    }
+    case "output": {
+      valueAttribute = SemanticConventions.OUTPUT_VALUE;
+      mimeTypeAttribute = SemanticConventions.OUTPUT_MIME_TYPE;
+      break;
+    }
+    default:
+      assertUnreachable(type);
+  }
+  if (io == null) {
+    return {};
+  }
+  const values = Object.values(io);
+  if (values.length === 1 && typeof values[0] === "string") {
+    return {
+      [valueAttribute]: values[0],
+      [mimeTypeAttribute]: MimeType.TEXT,
+    };
+  }
+
+  return {
+    [valueAttribute]: JSON.stringify(io),
+    [mimeTypeAttribute]: MimeType.JSON,
+  };
+}
+
+function getRoleFromMessageData(
+  messageData: Record<string, unknown>,
+): string | undefined {
+  const messageIds = messageData.lc_id;
+  if (!isNonEmptyArray(messageIds)) {
+    return;
+  }
+  const langchainMessageClass = messageIds[messageIds.length - 1];
+  const normalizedLangchainMessageClass = isString(langchainMessageClass)
+    ? langchainMessageClass.toLowerCase()
+    : "";
+  if (normalizedLangchainMessageClass.includes("human")) {
+    return "user";
+  }
+  if (normalizedLangchainMessageClass.includes("ai")) {
+    return "assistant";
+  }
+  if (normalizedLangchainMessageClass.includes("system")) {
+    return "system";
+  }
+  if (normalizedLangchainMessageClass.includes("function")) {
+    return "function";
+  }
+  if (
+    normalizedLangchainMessageClass.includes("chat") &&
+    isObject(messageData.kwargs) &&
+    isString(messageData.kwargs.role)
+  ) {
+    return messageData.kwargs.role;
+  }
+}
+
+function getContentFromMessageData(
+  messageKwargs: Record<string, unknown>,
+): string | undefined {
```

Comment:
any reason you are using undefined over null?

Reply:
updated to null